### PR TITLE
Describe discard statement

### PR DESCRIPTION
**Pull request by alan-baker**
* Add a description of discard works like GLSL discard (as opposed to
  HLSL discard)

### CODE REVIEW

File: wgsl/index.bs [hunk 1]
```diff
@@ -3621,7 +3621,15 @@ In this case the call site of this function invocation evaluates to the return v
...
 
-### Discard TODO ### {#discard-statement}
+### Discard ### {#discard-statement}
+
+The `discard` statement causes the current invocation to terminate and no
+entry point return value to be generated for the invocation. Only statements
+executed prior to the `discard` statement will have observable side effects. A
+`discard` statement may be executed by any function and the effect is the same,
+termination of the invocation occurs immediately upon executing the statement.
+After a `discard` statement is executed, control flow is non-uniform for the
```
Comment (kvark) [inline wgsl/index.bs:3631]: We may need to talk about this. Let's consider a case like this:
```rust
if (false) { discard; }
```
Technically it's never executed. Should we still consider the control flow to be non-uniform?

Or, more generally,
```rust
if (uniform_expression) {
  textureSample() // something that requires control flow being uniform
} else {
  discard; // makes the control flow non-uniform
}
```

If I understand correctly, this condition, as well as everything after it, is still in uniform control flow: either branch is taken by all invocations, and either all of them are discarded (in which case nothing past the condition matters anyway), or none of them.

**Review by kvark — APPROVED** [review]

File: wgsl/index.bs (hunk 1 above)
Comment (alan-baker) [reply wgsl/index.bs:3631]: For the first example, I phrased the spec such that the lose of uniformity only occurs if the discard is executed which I don't is the case.

For the second example, that will depend on the uniformity analysis and whether it counts the uniform expression as uniform or not.

**Comment by litherum** [comment]
> discard works like GLSL discard (as opposed to
HLSL discard)

so what are D3D implementations supposed to do?

**Comment by alan-baker** [comment]
It is easier to generate HLSL code that emulates the GLSL semantics than the or way around. So in the interest of portability I propose to use GLSL semantics. 

The are some subtle differences between the two semantics.

Code like the following *might* hang an implementation using D3D semantics, but would not hang GLSL semantics:
```
while (true) {
  // eventually every invocation sets cond to true.
  if (cond)
    discard;
}
```

HLSL semantics would allow correct derivatives in more cases, but if we discard as non-uniform then that is not a problem.

HLSL semantics are harder to pin down exactly when the helper invocations exist. Some implementations may start and stop them as necessary and all kill whole quads if they all become helpers, but between those points it difficult to describe. This will matter more if subgroups are exposed in the future.

In reality, without subgroups, they are generally interchangeable if disallow derivatives after discards and handle the loop cases. We could ban discard in loops for example (they are currently banned in continuing blocks) to make the interchange easier.

**Comment by dneto0** [comment]
> so what are D3D implementations supposed to do?

An easy translation is:

- set a private flag saying "discarded = true", then return
- consider this function as "may-discard"
- any call to a "may-discard" function must be followed by a   `if (discarded) return;` or `if (discarded) return nil;` where you make a nil-value of the return type of the current function.

**Comment by dneto0** [comment]
See #361 and #921

**Review by dneto0 — APPROVED** [review]

**Comment by kvark** [comment]
@dneto0 
> Or is the sticking point the fact that if all the invocations in a group 'discard' that they are all stopped executing and hence that is uniform? If that's the concern, I'm not concerned because calling the after-discarding state uniform or not uniform doesn't matter.

That's what I was wondering about. If we have `if (some_uniform_expression) { discard; }`, it will make all the invocations either pass or be discarded. So technically the control flow is still uniform after this, but the spec today would say that it's not.

**Comment by dneto0** [comment]
The  uniformity section hasn't been written yet. 

I suggest adding an "Issue" here to resolve this later.

**Comment by kvark** [comment]
@alan-baker I'm a bit confused now. So we are choosing GLSL semantics for discard() because it allows us to not consider the shadow invocations in the spec (and the issues associated with them), right? We still need to have this workaround on D3D.
*And* we still have shadow invocations, because pixels are shaded in 2x2 blocks anyway. So are we gaining anything from trying to hide the shadow invocation semantics? Or is the current direction chosen more because it's forward-compatible, in a sense that we can always say later that more places have uniform control flow, but not now?

**Comment by alan-baker** [comment]
GLSL semantics were chosen because it is easier to emulate HLSL semantics in terms of GLSL semantics than the other way around. It's not about hiding the existence of helper invocations. The spec will have to describe them. 

As for what can be declared to be uniform control flow, there will be a tension between satisfying developers (e.g. supporting dynamically uniform control flow) and bounding behaviour for security and portability reasons. I'm not certain how that will develop. At least if we start conservatively shaders will remain forward compatible as you say.

**Pull request merged**